File: 20240523-GrenlandWeb/TypeScript/src/UsersService.ts
import { Guid } from './Guid';
import { IDatabase } from './IDatabase';
import { IUsersService } from './IUsersService';
import { User } from './User';

export class UsersService extends IUsersService {

    constructor(readonly _database: IDatabase) {
        super();
    }

    async exists(username: string): Promise<boolean> {
        const user = await this._database.getCollectionFor(User).findOne({ username: username });
        return user !== null;
    }

    async register(username: string, password: string): Promise<Guid> {

        const id = Guid.create();
        await this._database.getCollectionFor(User).insertOne({ id, username, password } as any);
        return id;
    }
}

File: 20240523-GrenlandWeb/DotNET/SUT/UsersService.cs

using MongoDB.Driver;

namespace SUT;

public class UsersService(IDatabase database) : IUsersService
{
    readonly IDatabase _database = database;

    public async Task<bool> Exists(string userName) =>
        await _database.GetCollectionFor<User>().Find(u => u.UserName == userName).AnyAsync();

    public async Task<Guid> Register(string userName, string password)
    {
        var user = new User(Guid.NewGuid(), userName, password);
        await _database.GetCollectionFor<User>().InsertOneAsync(user);
        return user.Id;
    }
}
